ribbon.py

```diff
@@ -0,0 +1,545 @@
+            else:
+                ribbon = [ribbon_x, ribbon_y, ribbon_area, ribbon_contour]
+                self.follow_ribbon(ribbon, self.MARKER_SPEED)
+
+
+    def entry(self, marker_image, ribbon_image, image):
+        cropped_image = image[self.CROP_HEIGHT:(self.CROP_HEIGHT+45), 0:320]
+        blur_image = cv2.medianBlur(cropped_image, 3)
+        blur_image = cv2.cvtColor(blur_image, cv2.COLOR_RGB2HSV)
+        default_colour_bounds = ((40, 0, 0), (180, 255, 255))
+        limits = self.colour_bounds.get(
+            self.ALT_MARKER, default_colour_bounds
+        )
+        marker_mask = threshold_image(blur_image, limits)
+        marker_x, marker_y, marker_area, marker_contour = find_largest_contour(marker_mask)
+        pygame.mouse.set_pos(marker_y+self.CROP_HEIGHT, 320 - marker_x)
+        screen = pygame.display.get_surface()
+        frame = pygame.surfarray.make_surface(cv2.flip(marker_mask, 1))
+        screen.blit(frame, (self.CROP_HEIGHT, 0))
+        pygame.display.update()
+        if (marker_area > self.MIN_MARKER_AREA) and (marker_x > self.image_centre_x):  #>=go when marker on elft, <=go when marker on right
+            logger.info("Turntable barrier spotted at %d" % (marker_x))
+            self.drive.move(0, 1)
+            time.sleep(0.4)
+            self.drive.move(0, 0)
+
+    def leaving(self):
+        pass
+
+    def block_pushing(self):
+        pass
+
+    def direction(self, marker, ribbon):
+        '''function to check which side of the ribbon the tape marks are, indicating direction'''
+        if marker is not None and ribbon is not None:
+            image_offset = (320-self.MARKER_CROP_WIDTH)/2
+            marker_x = marker[0] + image_offset
+            ribbon_x = ribbon[0]
+            self.last_marker_time = time.time()
+            if (marker_x > ribbon_x) == self.MARKERS_ON_THE_LEFT:
+                 #if the markers are the same side as they're meant to be, we're going the right way
+                 direction = True
+            else:
+                 direction = False
+            if direction == self.MARKERS_ON_THE_LEFT:
+                logger.info ("marker spotted on the left")
+            else:
+                logger.info ("marker spotted on the right")
+        else:
+            #if either marker or ribbon can't be seen, assume we're ok
+            logger.info ("marker spotted but no ribbon")
+            direction = True
+        return direction
+
+    def stuck(self):
+        #if its been more than the timeout since we last saw a marker, we're probably stuck
+        if (self.last_marker_time + self.MARKER_TIMEOUT) < time.time() or ((self.last_t_error == self.last_before_that_t_error) and self.last_t_error is not None and self.last_t_error <> 0 and not self.isstuck):
+            self.isstuck = True
+        else:
+            self.isstuck = False
+        return False #self.isstuck
+
+    def escape(self):
+        print "escaping"
+        if random.choice([True, False]):
```

Comment:
`bool(random.getrandbits(1))` is a faster way of achieving this - see [helpful SO answer](https://stackoverflow.com/questions/6824681/get-a-random-boolean-in-python) for timings